### PR TITLE
Add bytes sent including headers to json results

### PR DESCRIPTION
**Pull request by olimpias**
Bytes that are sent to server,  are added to json result report under`sentRequestSize` and `sentRequestSizeBPS` label.

Ref:#185

### CODE REVIEW

**Comment by olimpias** [comment]
Hi @ldemailly, I think, https://fortio.istio.io:443 is down. Are you thinking to open a grpc server for integration tests on https://fortio.org since the project is not going under istio?

**Comment by ldemailly** [comment]
@olimpias thanks for noticing, I filled https://github.com/istio/istio/issues/10685

fortio is its own project but istio is using it and that site should be up/it's not expected it isn't

**Comment by olimpias** [comment]
Is there a way to trigger the CI pipeline?

**Comment by olimpias** [comment]
Somehow I was able to trigger build. @ldemailly  Are you able to see the problem on dockercloud from details? I'm having 404.

**Comment by ldemailly** [comment]
the failure was
```
=== RUN TestPingServer
Fortio 1.3.1-pre grpc 'foo' server listening on [::]:46300
Fortio 1.3.1-pre grpc 'foo' server listening on [::]:41779
12:25:33 I pingsrv.go:78> Using server certificate ../cert-tmp/server.crt to construct TLS credentials
12:25:33 I pingsrv.go:79> Using server key ../cert-tmp/server.key to construct TLS credentials
12:25:33 I pingsrv.go:151> Ping RTT 128737238 (avg of 124250675, 139507350, 122453690 ns) clock skew -53827229
12:25:34 I pingsrv.go:151> Ping RTT 156597907 (avg of 166259103, 166145674, 137388944 ns) clock skew -82499305
12:25:34 I pingsrv.go:151> Ping RTT 101864016 (avg of 101842115, 101903824, 101846109 ns) clock skew -50152981
12:25:34 I pingsrv.go:151> Ping RTT 118600815 (avg of 102594011, 126987633, 126220801 ns) clock skew -38405745
12:25:34 I pingsrv.go:151> Ping RTT 104749413 (avg of 102945639, 102895257, 108407343 ns) clock skew -50807698
12:25:34 I pingsrv.go:151> Ping RTT 104189465 (avg of 105079976, 105074644, 102413775 ns) clock skew -51844103
12:25:35 I pingsrv.go:151> Ping RTT 106401787 (avg of 108193863, 108203357, 102808141 ns) clock skew -53364839
Clock skew histogram usec : count 7 avg -54414.557 +/- 1.245e+04 min -82499.305 max -38405.745 sum -380901.9
# range, mid point, percentile, count
>= -82499.3 <= -38405.7 , -60452.5 , 100.00, 7
# target 50% -64127
RTT histogram usec : count 21 avg 117305.81 +/- 1.982e+04 min 101842.115 max 166259.103 sum 2463421.92
# range, mid point, percentile, count
>= 101842 <= 166259 , 134051 , 100.00, 21
# target 50% 132440
12:25:35 I grpcrunner.go:274> stripping https scheme. grpc destination: fortio.istio.io:443. grpc port: 443
12:25:35 E pingsrv.go:119> grpc error from Ping0 rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 35.197.102.163:443: connect: connection refused"
12:25:35 I grpcrunner.go:50> Using CA certificate ../cert-tmp/ca.crt to construct TLS credentials
12:25:35 I pingsrv.go:151> Ping RTT 1527647 (avg of 2471928, 1056434, 1054580 ns) clock skew 39714
12:25:35 I pingsrv.go:151> Ping RTT 1082574 (avg of 1133076, 1073613, 1041033 ns) clock skew 10770
12:25:35 I pingsrv.go:151> Ping RTT 1070567 (avg of 1079692, 1077791, 1054219 ns) clock skew 60201
12:25:35 I pingsrv.go:151> Ping RTT 1071996 (avg of 1131838, 1075801, 1008349 ns) clock skew -2290
12:25:35 I pingsrv.go:151> Ping RTT 17919891 (avg of 1065124, 26311239, 26383311 ns) clock skew 12683714
12:25:35 I pingsrv.go:151> Ping RTT 1066009 (avg of 1060871, 1046440, 1090718 ns) clock skew 33908
12:25:35 I pingsrv.go:151> Ping RTT 1075431 (avg of 1120358, 1053801, 1052135 ns) clock skew 8946
Clock skew histogram usec : count 7 avg 1833.5661 +/- 4430 min -2.29 max 12683.714 sum 12834.963
# range, mid point, percentile, count
>= -2.29 <= -2 , -2.145 , 14.29, 1
> 8 <= 10 , 9 , 28.57, 1
> 10 <= 12 , 11 , 42.86, 1
> 30 <= 40 , 35 , 71.43, 2
> 60 <= 70 , 65 , 85.71, 1
> 9990 <= 12683.7 , 11336.9 , 100.00, 1
# target 50% 32.5
RTT histogram usec : count 21 avg 3544.8739 +/- 7404 min 1008.349 max 26383.311 sum 74442.351
# range, mid point, percentile, count
>= 1008.35 <= 1200 , 1104.17 , 85.71, 18
> 2000 <= 2500 , 2250 , 90.48, 1
> 20000 <= 26383.3 , 23191.7 , 100.00, 2
# target 50% 1115.45
12:25:35 I grpcrunner.go:50> Using CA certificate ../cert-tmp/ca.crt to construct TLS credentials
12:25:35 E pingsrv.go:119> grpc error from Ping0 rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: tls: first record does not look like a TLS handshake"
12:25:35 E pingsrv.go:119> grpc error from Ping0 rpc error: code = Unavailable desc = transport is closing
RTT histogram usec : count 1 avg 3585.935 +/- 0 min 3585.935 max 3585.935 sum 3585.935
# range, mid point, percentile, count
>= 3585.93 <= 3585.93 , 3585.93 , 100.00, 1
# target 50% 3585.93
Health SERVING : 1
12:25:35 I grpcrunner.go:50> Using CA certificate ../cert-tmp/ca.crt to construct TLS credentials
RTT histogram usec : count 1 avg 95580.449 +/- 0 min 95580.449 max 95580.449 sum 95580.449
# range, mid point, percentile, count
>= 95580.4 <= 95580.4 , 95580.4 , 100.00, 1
# target 50% 95580.4
Health SERVING : 1
12:25:35 I grpcrunner.go:274> stripping https scheme. grpc destination: fortio.istio.io:443. grpc port: 443
12:25:35 E pingsrv.go:184> grpc error from Check rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 35.197.102.163:443: connect: connection refused"
RTT histogram usec : count 3 avg 5888.4747 +/- 6748 min 1073.513 max 15431.335 sum 17665.424
# range, mid point, percentile, count
>= 1073.51 <= 1200 , 1136.76 , 66.67, 2
> 10000 <= 15431.3 , 12715.7 , 100.00, 1
# target 50% 1136.76
Health SERVING : 3
12:25:35 I grpcrunner.go:50> Using CA certificate ../cert-tmp/ca.crt to construct TLS credentials
RTT histogram usec : count 3 avg 23623.064 +/- 3.189e+04 min 1008.242 max 68716.24799999999 sum 70869.191
# range, mid point, percentile, count
>= 1008.24 <= 1200 , 1104.12 , 66.67, 2
> 50000 <= 68716.2 , 59358.1 , 100.00, 1
# target 50% 1104.12
Health SERVING : 3
12:25:35 E pingsrv.go:184> grpc error from Check rpc error: code = NotFound desc = unknown service
12:25:35 I grpcrunner.go:50> Using CA certificate ../cert-tmp/ca.crt to construct TLS credentials
12:25:35 E pingsrv.go:184> grpc error from Check rpc error: code = NotFound desc = unknown service
12:25:35 E grpcrunner.go:47> Invalid TLS credentials: open ../missing/cert.crt: no such file or directory
12:25:35 C network.go:107> Can't listen to tcp socket 46300 (:46300) for grpc 'will fail': listen tcp :46300: bind: address already in use
--- FAIL: TestPingServer (2.37s)
pingsrv_test.go:38: insecure grpc ping server running, will connect to localhost:46300
pingsrv_test.go:41: secure grpc ping server running, will connect to localhost:41779
pingsrv_test.go:49: Unexpected result -1.000000, rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 35.197.102.163:443: connect: connection refused" with ping calls
pingsrv_test.go:71: Unexpected result <nil>, rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 35.197.102.163:443: connect: connection refused" with empty service health check
FAIL
FAIL	fortio.org/fortio/fgrpc	38.048s
12:25:41 I logger.go:97> Log level is now 0 Debug (was 2 Info)
```

I guess port issue, I clicked retry

File: fhttp/http_client.go [hunk 1]
```diff
@@ -311,6 +310,14 @@ func (c *Client) Close() int {
 	return 0 // TODO: find a way to track std client socket usage.
 }
 
+// GetRequestSize returns the message size
+func (c *Client) GetRequestSize() int {
+	if c.reqDump == nil {
```
Comment (ldemailly) [inline fhttp/http_client.go]: not needed
https://golang.org/pkg/builtin/#len

File: fhttp/http_client.go [hunk 2]
```diff
@@ -533,6 +547,14 @@ func (c *FastClient) returnRes() (int, []byte, int) {
 	return c.code, c.buffer[:c.size], c.headerLen
 }
 
+// GetRequestSize returns the message size
+func (c *FastClient) GetRequestSize() int {
+	if c.req == nil {
```
Comment (ldemailly) [inline fhttp/http_client.go]: same

File: fhttp/httprunner.go [hunk 3]
```diff
@@ -40,10 +40,12 @@ type HTTPRunnerResults struct {
...
 	headerSizes *stats.Histogram
 	// exported result
-	Sizes       *stats.HistogramData
-	HeaderSizes *stats.HistogramData
-	URL         string
-	SocketCount int
+	Sizes              *stats.HistogramData
+	HeaderSizes        *stats.HistogramData
+	URL                string
+	SentRequestSize    int
```
Comment (ldemailly) [inline fhttp/httprunner.go]: add comments above the new fields and that would also break up the auto indent into multiple chunks making the diff smaller/easier to review

File: fhttp/httprunner.go [hunk 4]
```diff
@@ -154,6 +158,7 @@ func RunHTTPTest(o *HTTPRunnerOptions) (*HTTPRunnerResults, error) {
 		total.sizes.Transfer(httpstate[i].sizes)
 		total.headerSizes.Transfer(httpstate[i].headerSizes)
 	}
+	total.SentRequestSizeBPS = float64(total.SentRequestSize) / total.ActualDuration.Seconds()
```
Comment (ldemailly) [inline fhttp/httprunner.go]: why is that needed to be persisted ? also BPS is ambigous

**Review by ldemailly — CHANGES_REQUESTED** [review]
thanks for the pr, happy new year! see below:

File: fhttp/http_client.go (hunk 1 above)
Comment (olimpias) [reply fhttp/http_client.go]: My purpose was to get request size from one method for two different structs otherwise I need to cast the interface to `FastClient` or `Client` to get request size. Whats your recommendation here? Should I cast the interface to `FastClient` or `Client` to get request size in `Run` method?

File: fhttp/httprunner.go (hunk 3 above)
Comment (olimpias) [reply fhttp/httprunner.go]: Sure

File: fhttp/httprunner.go (hunk 4 above)
Comment (olimpias) [reply fhttp/httprunner.go]: At the #185 task you told me to collect requestSize and in terms of kbps.

> we should add send size and kbytes/sec calc

BPS was my fault. I had to make it kbps instead of bps. But if you are saying that we shouldnt persist the Kbps value as well, it is ok for me.

**Comment by olimpias** [comment]
Happy new year!

File: fhttp/http_client.go (hunk 1 above)
Comment (ldemailly) [reply fhttp/http_client.go]: I am just pointing out that you don't need to test if reqDump is `nil` you can replace by just line 318

`return len(c.reqDump)`

File: fhttp/httprunner.go (hunk 3 above)
Comment (ldemailly) [reply fhttp/httprunner.go]: ping :)

File: fhttp/httprunner.go (hunk 4 above)
Comment (ldemailly) [reply fhttp/httprunner.go]: Back in that task I was thinking of changing the text summary output from:
```
All done 40 calls (plus 4 warmup) 40.771 ms avg, 7.9 qps
```
to maybe
```
All done 40 calls (plus 4 warmup) 40.771 ms avg, 7.9 qps, 23.2 kB/s received, 0.7 kB/s sent.
```

File: fhttp/httprunner.go (hunk 3 above)
Comment (olimpias) [reply fhttp/httprunner.go]: I was waiting for your comments :D.

File: fhttp/httprunner.go (hunk 4 above)
Comment (olimpias) [reply fhttp/httprunner.go]: Sure, I can make it like that.

**Comment by olimpias** [comment]
I added one TODO. I will work on it after this is merged. (I m not sure, if we are able to receive the message size that we sent to server in the GRPC lib or response size from a server. I will take a look...)

File: fhttp/httprunner.go [hunk 5]
```diff
@@ -43,6 +43,11 @@ type HTTPRunnerResults struct {
 	Sizes       *stats.HistogramData
 	HeaderSizes *stats.HistogramData
 	URL         string
+	// total message size that sent to server
+	SentRequestSize int
+	// total message size that received from the server
+	ReceivedResponseSize int
```
Comment (ldemailly) [inline fhttp/httprunner.go]: This already exist as the sizes histogram

File: fhttp/httprunner.go [hunk 6]
```diff
@@ -43,6 +43,11 @@ type HTTPRunnerResults struct {
 	Sizes       *stats.HistogramData
 	HeaderSizes *stats.HistogramData
 	URL         string
+	// total message size that sent to server
+	SentRequestSize int
```
Comment (ldemailly) [inline fhttp/httprunner.go]: type should probably be uint64 (or same type as histogram/count sum)

File: periodic/periodic.go [hunk 7]
```diff
@@ -575,3 +579,34 @@ func (r *RunnerResults) ID() string {
...
+			r.ActualQPS)
+	}
+	return fmt.Sprintf("All done %d calls %.3f ms avg, %.1f qps, %.1f kB/s received, %.1f kB/s sent.\n",
+		r.DurationHistogram.Count,
+		1000.*r.DurationHistogram.Avg,
+		r.ActualQPS, r.ReceivedResponseSizeKPBS, r.SentRequestSizeKBPS)
+}
+
+// StringWithWarmUp prints the Runner Results with Warmup
+func (r *RunnerResults) StringWithWarmUp(warmup int) string {
```
Comment (ldemailly) [inline periodic/periodic.go]: call this InfoTextWithWarmup()

File: periodic/periodic.go [hunk 8]
```diff
@@ -575,3 +579,34 @@ func (r *RunnerResults) ID() string {
 	}
 	return base
 }
+
+//TODO remove above if condition when receive and sent byte count is added to GRPC as well.
+// String prints the Runner Results
+func (r *RunnerResults) String() string {
+	if r.SentRequestSizeKBPS < 0.01 && r.ReceivedResponseSizeKPBS < 0.01 {
+		return fmt.Sprintf("All done %d calls %.3f ms avg, %.1f qps,\n",
```
Comment (ldemailly) [inline periodic/periodic.go]: coma and \n is probably wrong

also just add to the common string instead of duplicating the whole common part

File: periodic/periodic.go [hunk 9]
```diff
@@ -575,3 +579,34 @@ func (r *RunnerResults) ID() string {
 	}
 	return base
 }
+
+//TODO remove above if condition when receive and sent byte count is added to GRPC as well.
```
Comment (ldemailly) [inline periodic/periodic.go]: I think you mean below, not above

Comment (ldemailly) [reply periodic/periodic.go]: you don’t want to update the comment ?

File: fhttp/httprunner.go [hunk 10]
```diff
@@ -154,6 +159,8 @@ func RunHTTPTest(o *HTTPRunnerOptions) (*HTTPRunnerResults, error) {
 		total.sizes.Transfer(httpstate[i].sizes)
 		total.headerSizes.Transfer(httpstate[i].headerSizes)
 	}
+	total.SentRequestSizeKBPS = float64(total.SentRequestSize) / (total.ActualDuration.Seconds() * 1000)
+	total.ReceivedResponseSizeKPBS = (total.sizes.Sum + total.headerSizes.Sum) / (total.ActualDuration.Seconds() * 1000)
```
Comment (ldemailly) [inline fhttp/httprunner.go]: not KP but KB

also if go lint allows let’s use ...SizeKBperSec

File: periodic/periodic.go (hunk 9 above)
Comment (olimpias) [reply periodic/periodic.go]: Actually, I changed it but there was a problem with todo. Lint was asking for a empty line between method comment and todo. I did revert. After that I forgot to update it back. Sorry.

**Review by ldemailly — COMMENTED** [review]
Thanks! can we keep the test coverage up?

File: fhttp/http_client.go [hunk 11]
```diff
@@ -276,19 +279,14 @@ func newHTTPRequest(o *HTTPOptions) *http.Request {
 	if !log.LogDebug() {
```
Comment (ldemailly) [inline fhttp/http_client.go]: this if is useless now, please remove

File: fortio_main.go [hunk 12]
```diff
@@ -315,11 +315,7 @@ func fortioLoad(justCurl bool, percList []float64) {
 	if ro.Exactly > 0 {
 		warmup = 0
 	}
-	_, _ = fmt.Fprintf(out, "All done %d calls (plus %d warmup) %.3f ms avg, %.1f qps\n",
-		rr.DurationHistogram.Count,
-		warmup,
-		1000.*rr.DurationHistogram.Avg,
-		rr.ActualQPS)
+	_, _ = fmt.Fprintln(out, rr.InfoTextWithWarmup(warmup))
```
Comment (ldemailly) [inline fortio_main.go:320]: let's just move the code into the runner completely - warmup or not is determined by the runner based on "Exactly" anyway and is (or should have been) the same anyway between fortio_main and uihandler

File: periodic/periodic.go [hunk 13]
```diff
@@ -442,7 +452,7 @@ func (r *periodicRunner) Run() RunnerResults {
 		requestedDuration += fmt.Sprintf(", interrupted after %d", actualCount)
 	}
 	result := RunnerResults{r.RunType, r.Labels, start, requestedQPS, requestedDuration,
-		actualQPS, elapsed, r.NumThreads, version.Short(), functionDuration.Export().CalcPercentiles(r.Percentiles), r.Exactly}
+		actualQPS, elapsed, r.NumThreads, version.Short(), functionDuration.Export().CalcPercentiles(r.Percentiles), r.Exactly, 0, 0}
```
Comment (ldemailly) [inline periodic/periodic.go]: if the field is in RunnerResult (as opposed to Http or Grpc specific) it should be set here

**Review by ldemailly — COMMENTED** [review]
This is going to take a bit more work to do right - it has to do with the Runner vs Http/GRPC subclass

I can do it if you prefer / I will look a bit later

**Comment by olimpias** [comment]
`InfoTextWithWarmup` and `InfoText` methods are reducing the test coverage. At evening, I can add GRPC as well and remove the if condition from the methods also add their test methods. For your the latest comment, it is up to you. I can look into it, if you want :).

**Comment by olimpias** [comment]
I added GRPC as well but I m not sure, if I added GRPC receive and sent size KBPS in a right way. I'm leaving you the two refactoring things, probably you have core changes in your mind.

File: fgrpc/grpcrunner.go [hunk 14]
```diff
@@ -83,6 +86,23 @@ type GRPCRunnerResults struct {
...
+
+// GetRequestSize returns the request body size. If the request is Ping, Ping Message size will return. Otherwise
+// HealthCheckRequest's service len will return.
+func (grpcstate *GRPCRunnerResults) GetRequestSize() int {
+	if grpcstate.Ping {
+		size := int(unsafe.Sizeof(&grpcstate.reqP))
+		size += len(grpcstate.reqP.Payload)
+		return size
+	}
+	return len(grpcstate.reqH.Service)
```
Comment (ldemailly) [inline fgrpc/grpcrunner.go:105]: I doubt that's the wire size of the packet sent, there is probably some header, method name or id, etc

also there should be a way to get that request size or bytes from grpc ? did you search for it ?

File: fgrpc/grpcrunner.go [hunk 15]
```diff
@@ -92,16 +112,25 @@ func (grpcstate *GRPCRunnerResults) Run(t int) {
...
-		r, err = grpcstate.clientH.Check(context.Background(), &grpcstate.reqH)
+		r, err = grpcstate.clientH.Check(ctx, &grpcstate.reqH)
 		if r != nil {
 			status = r.Status
 			res = r
 		}
 	}
+	headers, ok := metadata.FromIncomingContext(ctx)
+	if ok {
+		grpcstate.SentMessageHeaderSize += uint64(headers.Len())
```
Comment (ldemailly) [inline fgrpc/grpcrunner.go:128]: that's more like it ?

**Comment by ldemailly** [comment]
I have been traveling and visiting/taking care of family so my bandwidth is limited, will look more when I can (soon)

File: fgrpc/grpcrunner.go (hunk 14 above)
Comment (olimpias) [reply fgrpc/grpcrunner.go:105]: I didn't have time to look in a detailed way. But I will hopefully take a look in the weekend.

**Comment by ldemailly** [comment]
Sorry for the delays, can you merge/resolve conflicts ?

**Comment by olimpias** [comment]
I accidentally pushed my requestSize branch to fortio, sorry. If you are okay, we can merge it without GRPC. I found a way to get http Headers from GRPC however I couldn't find a way to get request headers from GRPC. I stuck with this problem.

**Comment by ldemailly** [comment]
closing this for now but do reopen if you want
